File: frontend/src/modules/forms/conditions.js
// Conditional logic: which questions apply, given the answers so far.
//
// A mirror of app/modules/forms/conditions.py — same operators, same targets,
// same fallbacks. The form reacts as it is filled in without asking the server
// anything; the server evaluates the identical rules again when the answers
// arrive, because a request can be sent without ever opening the form.
//
// Two rules run through it. Nothing is parsed and nothing is executed: a
// condition is three values and an operator looked up in a table. And nothing
// ever mentions a label — a rule names the key an answer is stored under and
// compares against the stored value, so translating the form or a catalogue
// changes what is read and never what is compared.

import { fieldHidden } from './fieldTypes.js'

const text = (value) => (value == null ? '' : String(value).trim())

const empty = (value) =>
  value == null || value === '' ||
  (Array.isArray(value) && value.length === 0) ||
  (typeof value === 'object' && !Array.isArray(value) && Object.keys(value).length === 0)

const truth = (value) =>
  typeof value === 'boolean' ? value : ['true', 'yes', 'y', '1'].includes(text(value).toLowerCase())

/**
 * Whether an answer is the value a rule names.
 *
 * Compared as text, because a form holds "18" where the definition says 18 and
 * a rule written either way must mean the same thing. A multi-select answer
 * matches when the value is among the ones chosen.
 */
const same = (answer, wanted) => {
  if (Array.isArray(answer)) return answer.some((item) => text(item) === text(wanted))
  if (typeof answer === 'boolean' || typeof wanted === 'boolean') {
    return truth(answer) === truth(wanted)
  }
  return text(answer) === text(wanted)
}

const number = (value) => {
  const parsed = Number(text(value))
  return text(value) !== '' && Number.isFinite(parsed) ? parsed : null
}

/** A numeric comparison, or false when either side is not a number. */
const compare = (answer, wanted, decide) => {
  const left = number(answer)
  const right = number(wanted)
  return left !== null && right !== null && decide(left, right)
}

const contains = (answer, wanted) => {
  if (Array.isArray(answer)) return answer.some((item) => text(item) === text(wanted))
  return text(answer).toLowerCase().includes(text(wanted).toLowerCase())
}

/** The operators a rule may use. A table, so adding one is a line here. */
export const OPERATORS = {
  equals: same,
  not_equals: (a, w) => !same(a, w),
  is_empty: (a) => empty(a),
  is_not_empty: (a) => !empty(a),
  greater_than: (a, w) => compare(a, w, (x, y) => x > y),
  greater_than_or_equal: (a, w) => compare(a, w, (x, y) => x >= y),
  less_than: (a, w) => compare(a, w, (x, y) => x < y),
  less_than_or_equal: (a, w) => compare(a, w, (x, y) => x <= y),
  contains,
  not_contains: (a, w) => !contains(a, w),
}

/** How each operator reads in the builder, and whether it compares to anything. */
export const OPERATOR_LABELS = [
  ['equals', 'is'],
  ['not_equals', 'is not'],
  ['is_empty', 'is blank'],
  ['is_not_empty', 'is answered'],
  ['greater_than', 'is more than'],
  ['greater_than_or_equal', 'is at least'],
  ['less_than', 'is less than'],
  ['less_than_or_equal', 'is at most'],
  ['contains', 'includes'],
  ['not_contains', 'does not include'],
]

export const UNARY = ['is_empty', 'is_not_empty']

/** One condition against the answers so far. */
export function evaluate(condition, answers) {
  if (!condition) return false
  const decide = OPERATORS[text(condition.operator)]
  // An unknown operator is false, never an error: a definition written by a
  // newer version must not make this one unusable.
  if (!decide) return false
  return Boolean(decide((answers || {})[text(condition.field)], condition.value))
}

/** Whether a rule's conditions hold. A rule with no conditions never fires. */
export function evaluateRule(rule, answers) {
  const conditions = (rule?.conditions || []).filter(Boolean)
  if (!conditions.length) return false

  return text(rule.logic).toUpperCase() === 'OR'
    ? conditions.some((c) => evaluate(c, answers))
    : conditions.every((c) => evaluate(c, answers))
}

export const evaluateRules = (rules, answers) =>
  (rules || []).map((rule) => evaluateRule(rule, answers))

/** Every field a rule reads. These stay answerable whatever the rules say. */
export function controllingFields(rules) {
  const names = new Set()
  for (const rule of rules || []) {
    for (const condition of rule?.conditions || []) {
      if (condition?.field) names.add(text(condition.field))
    }
  }
  return names
}

/**
 * Which parts of the form do not apply, given these answers.
 *
 * Everything is visible until a rule says otherwise, so a form with no rules is
 * unaffected. A `show` rule whose conditions do not hold hides its target; a
 * `hide` rule whose conditions do hold hides it.
 *
 * A field a rule reads is never hidden by it — the question controlling the
 * questionnaire has to stay answerable, or nothing could ever be shown again.
 */
export function hidden(formJson, answers) {
  const rules = (formJson?.rules || []).filter(Boolean)

  const fields = new Set()
  const sections = new Set()
  let form = false

  for (const rule of rules) {
    const holds = evaluateRule(rule, answers)
    const action = text(rule.action).toLowerCase() || 'show'
    if ((action === 'show') === holds) continue

    const target = rule.target || {}
    const kind = text(target.type).toLowerCase()

    if (kind === 'field' && target.name) fields.add(text(target.name))
    else if ((kind === 'section' || kind === 'group') && (target.key || target.name)) {
      sections.add(text(target.key || target.name))
    } else if (kind === 'form') form = true
  }

  for (const field of formJson?.fields || []) {
    if (!field?.name) continue
    if (form || (field.section && sections.has(field.section))) fields.add(field.name)
  }

  for (const name of controllingFields(rules)) fields.delete(name)

  // A question hidden outright (`config.hide`) is hidden whatever the rules
  // say — after the exception above, not before it: a question nobody can see
  // cannot be the one that makes a rule hold. The backend's conditions.py does
  // the same, in the same order.
  for (const field of formJson?.fields || []) {
    if (field?.name && fieldHidden(field)) fields.add(field.name)
  }

  return { fields, sections, form }
}

/**
 * The answers that currently apply.
 *
 * What gets submitted. Values for hidden questions stay in the page's state, so
 * changing the controlling answer back brings them straight back — but they are
 * left out of the payload, because the form is not asking them. The server
 * refuses them if they are sent anyway.
 */
export function applicable(formJson, values) {
  const { fields } = hidden(formJson, values)
  if (!fields.size) return values

  const kept = {}
  for (const [name, value] of Object.entries(values || {})) {
    if (!fields.has(name)) kept[name] = value
  }
  return kept
}


/* ── a question that is renamed or deleted ──────────────────────────────────
 *
 * A rule refers to questions by name: the one it controls (`target`) and the
 * ones it reads (`conditions[].field`). The builder already carries a rename
 * into the layout and the WhatsApp conversation, and drops both when a question
 * goes — but rules were left behind, so deleting a question left a rule about a
 * question that no longer existed and the server refused the whole form:
 *
 *     rules.0.target: This rule controls 'gender', which is not a question
 *     on this form.
 *
 * From then on nothing could be saved. The form on screen and the form in the
 * database drifted apart, and a reload looked like lost work.
 */

/** Is this rule's target the named question? Sections and the form are not. */
const controls = (rule, name) =>
  rule?.target?.type === 'field' && text(rule.target.name) === text(name)

/**
 * The rules, with one question's name changed everywhere it appears.
 *
 * Returns the same array when nothing refers to it, so a caller can tell
 * cheaply that there was nothing to do.
 */
export function renameFieldInRules(rules, from, to) {
  if (!Array.isArray(rules) || !from || !to || from === to) return rules

  let changed = false

  const next = rules.map((rule) => {
    if (!rule) return rule

    const conditions = (rule.conditions || []).map((condition) => {
      if (text(condition?.field) !== text(from)) return condition
      changed = true
      return { ...condition, field: to }
    })

    if (!controls(rule, from)) {
      return changed ? { ...rule, conditions } : rule
    }

    changed = true
    return { ...rule, conditions, target: { ...rule.target, name: to } }
  })

  return changed ? next : rules
}

/**
 * The rules, with every reference to one question removed.
 *
 * A rule that controlled it goes entirely — there is nothing left to show or
 * hide. A rule that merely read it loses that condition, and goes too if it was
 * the only one: a rule with nothing to test would never fire, and its target
 * would disappear for good. That is the same reasoning `ConditionEditor` uses
 * when the last condition is removed by hand.
 */
export function removeFieldFromRules(rules, name) {
  if (!Array.isArray(rules) || !name) return rules

  let changed = false

  const next = []
  for (const rule of rules) {
    if (!rule) continue

    if (controls(rule, name)) {
      changed = true
      continue
    }

    const conditions = (rule.conditions || [])
      .filter((condition) => text(condition?.field) !== text(name))

    if (conditions.length === (rule.conditions || []).length) {
      next.push(rule)
      continue
    }

    changed = true
    if (conditions.length) next.push({ ...rule, conditions })
  }

  return changed ? next : rules
}

File: backend/app/modules/forms/conditions.py
"""Conditional logic: which questions apply, given the answers so far.

A form often asks something only when an earlier answer calls for it. Rather
than a special case per form, the definition carries rules:

    {"rules": [
      {"conditions": [{"field": "consent", "operator": "equals", "value": "yes"}],
       "logic": "AND",
       "action": "show",
       "target": {"type": "form"}}
    ]}

Three things this deliberately is not.

**It is not an expression language.** No string is parsed and nothing is ever
executed — a condition is three values and an operator looked up in a table, so
a definition can be edited by anyone without becoming a way to run code.

**It never mentions a label.** A rule names a field by the key answers are
stored under and compares against the stored value — `consent == "yes"`, not
`"Consent" == "Yes"`. That is what lets the same form work in every language
and with every catalogue: translating a form or a catalogue changes what is
read, never what is compared. See `translations.py` and
`client_catalog/catalog_options.py`, which both hold the value fixed and vary
only the label.

**It is not the frontend's alone.** These same semantics are mirrored in
`frontend/src/modules/forms/conditions.js` so a form reacts as it is filled in,
but the submission service evaluates them again on arrival. A hidden question is
not an optional one: an answer to something the form did not ask is refused.

Targets are the structures the form already has:

    {"type": "field", "name": "..."}     one question
    {"type": "section", "key": "..."}    every question in that section
    {"type": "form"}                     the whole questionnaire

`group` is accepted as another word for `section` — the section *is* the group
of questions here, and adding a parallel structure to mean the same thing would
be two things to keep in step.
"""
import logging
from typing import Any, Dict, List, Optional, Set

logger = logging.getLogger(__name__)


# --------------------------------------------------------------------------- #
# operators
# --------------------------------------------------------------------------- #
def _text(value: Any) -> str:
    return "" if value is None else str(value).strip()


def _empty(value: Any) -> bool:
    return value is None or value == "" or value == [] or value == {}


def _same(answer: Any, wanted: Any) -> bool:
    """Whether an answer is the value a rule names.

    Compared as text, because a form sends `"18"` where the definition says
    `18`, and a rule written either way must mean the same thing. A multi-select
    answer matches when the value is among the ones chosen.
    """
    if isinstance(answer, list):
        return any(_text(item) == _text(wanted) for item in answer)
    if isinstance(answer, bool) or isinstance(wanted, bool):
        return _truth(answer) == _truth(wanted)
    return _text(answer) == _text(wanted)


def _truth(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    return _text(value).lower() in ("true", "yes", "y", "1")


def _number(value: Any) -> Optional[float]:
    try:
        return float(_text(value))
    except (TypeError, ValueError):
        return None


def _compare(answer: Any, wanted: Any, decide) -> bool:
    """A numeric comparison, or False when either side is not a number.

    Not an error: a rule comparing a text answer to a number simply does not
    hold, and a half-filled form must not be unanswerable because of it.
    """
    left, right = _number(answer), _number(wanted)
    if left is None or right is None:
        return False
    return decide(left, right)


def _contains(answer: Any, wanted: Any) -> bool:
    """Whether the answer includes the value.

    A multi-select holds a list, so membership; anything else is text, so a
    substring — which is what "contains" means for a written answer.
    """
    if isinstance(answer, list):
        return any(_text(item) == _text(wanted) for item in answer)
    return _text(wanted).lower() in _text(answer).lower()


# The operators a rule may use, by name. A table rather than a chain of ifs, so
# adding one is a line here and a line in the frontend's mirror — and so nothing
# outside this module ever has to know how a comparison is made.
OPERATORS = {
    "equals": _same,
    "not_equals": lambda a, w: not _same(a, w),
    "is_empty": lambda a, w: _empty(a),
    "is_not_empty": lambda a, w: not _empty(a),
    "greater_than": lambda a, w: _compare(a, w, lambda x, y: x > y),
    "greater_than_or_equal": lambda a, w: _compare(a, w, lambda x, y: x >= y),
    "less_than": lambda a, w: _compare(a, w, lambda x, y: x < y),
    "less_than_or_equal": lambda a, w: _compare(a, w, lambda x, y: x <= y),
    "contains": _contains,
    "not_contains": lambda a, w: not _contains(a, w),
}

# The operators that compare against nothing — their `value` is meaningless and
# the builder hides the box for it.
UNARY = ("is_empty", "is_not_empty")

ACTIONS = ("show", "hide")
LOGIC = ("AND", "OR")
TARGET_TYPES = ("field", "section", "form")


# --------------------------------------------------------------------------- #
# evaluating
# --------------------------------------------------------------------------- #
def evaluate(condition: Dict[str, Any], answers: Dict[str, Any]) -> bool:
    """One condition against the answers so far.

    An unknown operator is False rather than an error: a definition written by a
    newer version of this application must not make an old one unusable.
    """
    if not isinstance(condition, dict):
        return False

    decide = OPERATORS.get(_text(condition.get("operator")))
    if decide is None:
        logger.warning("Unknown condition operator %r", condition.get("operator"))
        return False

    return bool(decide(answers.get(_text(condition.get("field"))), condition.get("value")))


def evaluate_rule(rule: Dict[str, Any], answers: Dict[str, Any]) -> bool:
    """Whether a rule's conditions hold. A rule with no conditions never fires."""
    if not isinstance(rule, dict):
        return False

    conditions = [c for c in (rule.get("conditions") or []) if isinstance(c, dict)]
    if not conditions:
        return False

    results = (evaluate(condition, answers) for condition in conditions)
    return any(results) if _text(rule.get("logic")).upper() == "OR" else all(results)


def evaluate_rules(rules: List[Dict[str, Any]], answers: Dict[str, Any]) -> List[bool]:
    """Each rule's verdict, in order. For a caller that wants to explain itself."""
    return [evaluate_rule(rule, answers) for rule in rules or []]


# --------------------------------------------------------------------------- #
# what that means for a form
# --------------------------------------------------------------------------- #
def _targets(rule: Dict[str, Any]) -> Dict[str, str]:
    target = rule.get("target")
    return target if isinstance(target, dict) else {}


def hidden(form_json: Dict[str, Any], answers: Dict[str, Any]) -> Dict[str, Any]:
    """Which parts of the form do not apply, given these answers.

    Returns the hidden field names, the hidden section keys, and whether the
    questionnaire as a whole is hidden.

    Everything is visible until a rule says otherwise, so a form with no rules —
    every form built before this existed — is unaffected. A `show` rule whose
    conditions do not hold hides its target; a `hide` rule whose conditions do
    hold hides it. Several rules on one target are read together: if any of them
    hides it, it is hidden.

    A field a rule reads is never hidden by that rule. The question controlling
    the questionnaire has to stay answerable, or nothing could ever be shown.
    """
    from app.modules.forms.form_schema import field_name

    rules = [r for r in (form_json.get("rules") or []) if isinstance(r, dict)]

    hidden_fields: Set[str] = set()
    hidden_sections: Set[str] = set()
    form_hidden = False

    for rule in rules:
        holds = evaluate_rule(rule, answers)
        action = _text(rule.get("action")).lower() or "show"

        # show + conditions unmet, or hide + conditions met.
        if (action == "show") == holds:
            continue

        target = _targets(rule)
        kind = _text(target.get("type")).lower()

        if kind == "field":
            name = _text(target.get("name"))
            if name:
                hidden_fields.add(name)
        elif kind in ("section", "group"):
            key = _text(target.get("key") or target.get("name"))
            if key:
                hidden_sections.add(key)
        elif kind == "form":
            form_hidden = True

    # Everything in a hidden section is hidden with it, and everything at all is
    # hidden when the questionnaire is.
    for field in form_json.get("fields") or []:
        if not isinstance(field, dict):
            continue
        name = field_name(field)
        if not name:
            continue
        if form_hidden or (field.get("section") and field["section"] in hidden_sections):
            hidden_fields.add(name)

    # …except the questions the rules read. A form-level rule that hid its own
    # controlling question could never be satisfied again.
    for name in controlling_fields(rules):
        hidden_fields.discard(name)

    return {
        "fields": sorted(hidden_fields),
        "sections": sorted(hidden_sections),
        "form": form_hidden,
    }


def controlling_fields(rules: List[Dict[str, Any]]) -> Set[str]:
    """Every field a rule reads. These stay answerable whatever the rules say."""
    names: Set[str] = set()
    for rule in rules or []:
        if not isinstance(rule, dict):
            continue
        for condition in rule.get("conditions") or []:
            if isinstance(condition, dict):
                name = _text(condition.get("field"))
                if name:
                    names.add(name)
    return names


def applies(form_json: Dict[str, Any], answers: Dict[str, Any], name: str) -> bool:
    """Whether one question applies. Convenience over `hidden`."""
    return name not in set(hidden(form_json, answers)["fields"])


# --------------------------------------------------------------------------- #
# cleaning what arrives
# --------------------------------------------------------------------------- #
def normalize_rules(raw: Any) -> List[Dict[str, Any]]:
    """Keep only rules this engine can actually evaluate.

    Called from `normalize_form`, so a hand-edited definition cannot store a
    shape the renderer would choke on. Anything unrecognised is dropped rather
    than rejected: a bad rule must never stop a form being saved, and a dropped
    rule leaves its target visible, which is the safe direction to fail.
    """
    if not isinstance(raw, list):
        return []

    cleaned: List[Dict[str, Any]] = []

    for item in raw:
        if not isinstance(item, dict):
            continue

        conditions = []
        for condition in item.get("conditions") or []:
            if not isinstance(condition, dict):
                continue
            field = _text(condition.get("field"))
            operator = _text(condition.get("operator"))
            if not field or operator not in OPERATORS:
                continue

            kept = {"field": field, "operator": operator}
            if operator not in UNARY:
                kept["value"] = condition.get("value")
            conditions.append(kept)

        if not conditions:
            continue

        target = item.get("target")
        if not isinstance(target, dict):
            continue

        kind = _text(target.get("type")).lower()
        # One word for one thing: a group of questions is a section here.
        if kind == "group":
            kind = "section"
        if kind not in TARGET_TYPES:
            continue

        cleaned_target: Dict[str, Any] = {"type": kind}
        if kind == "field":
            name = _text(target.get("name"))
            if not name:
                continue
            cleaned_target["name"] = name
        elif kind == "section":
            key = _text(target.get("key") or target.get("name"))
            if not key:
                continue
            cleaned_target["key"] = key

        action = _text(item.get("action")).lower()
        logic = _text(item.get("logic")).upper()

        cleaned.append({
            "conditions": conditions,
            "logic": logic if logic in LOGIC else "AND",
            "action": action if action in ACTIONS else "show",
            "target": cleaned_target,
        })

    return cleaned


def problems(form_json: Dict[str, Any]) -> List[Dict[str, str]]:
    """Configurations that cannot work, for the validation pipeline.

    Three of them. A question cannot depend on itself — its own answer decides
    whether it may be answered. A rule cannot read a field the form does not
    have. And a chain of rules cannot come back round to where it started, since
    nothing would settle.
    """
    from app.modules.forms.form_schema import field_name

    rules = form_json.get("rules") or []
    names = {field_name(f) for f in form_json.get("fields") or []
             if isinstance(f, dict) and field_name(f)}
    sections = {s.get("key") for s in form_json.get("sections") or []
                if isinstance(s, dict)}

    found: List[Dict[str, str]] = []

    # Which fields each field's visibility depends on, for the cycle check.
    depends: Dict[str, Set[str]] = {}

    for index, rule in enumerate(rules):
        target = _targets(rule)
        kind = _text(target.get("type"))
        reads = {_text(c.get("field")) for c in rule.get("conditions") or []
                 if isinstance(c, dict)}

        for name in reads:
            if name and name not in names:
                found.append({
                    "path": f"rules.{index}.conditions",
                    "message": f"This rule asks about '{name}', which is not a question "
                               f"on this form.",
                })

        if kind == "field":
            name = _text(target.get("name"))
            if name and name not in names:
                found.append({
                    "path": f"rules.{index}.target",
                    "message": f"This rule controls '{name}', which is not a question "
                               f"on this form.",
                })
            if name in reads:
                found.append({
                    "path": f"rules.{index}",
                    "message": f"'{name}' cannot decide whether it is asked — a question "
                               f"cannot depend on its own answer.",
                })
            if name:
                depends.setdefault(name, set()).update(reads - {name})

        elif kind == "section":
            key = _text(target.get("key"))
            if key and key not in sections:
                found.append({
                    "path": f"rules.{index}.target",
                    "message": f"This rule controls the section '{key}', which this form "
                               f"does not have.",
                })
            # Everything in the section inherits the dependency.
            for field in form_json.get("fields") or []:
                if isinstance(field, dict) and field.get("section") == key:
                    name = field_name(field)
                    if name in reads:
                        found.append({
                            "path": f"rules.{index}",
                            "message": f"'{name}' is in the section this rule controls and "
                                       f"is also what it asks about, so it would decide "
                                       f"whether it is asked.",
                        })
                    if name:
                        depends.setdefault(name, set()).update(reads - {name})

    for name in sorted(_cycles(depends)):
        found.append({
            "path": "rules",
            "message": f"The rules around '{name}' depend on each other in a circle, so "
                       f"nothing would decide whether it is asked.",
        })

    return found


def _cycles(depends: Dict[str, Set[str]]) -> Set[str]:
    """The names caught in a dependency circle."""
    caught: Set[str] = set()

    for start in depends:
        seen: Set[str] = set()
        stack = [start]
        while stack:
            name = stack.pop()
            for other in depends.get(name, ()):
                if other == start:
                    caught.add(start)
                    stack = []
                    break
                if other not in seen:
                    seen.add(other)
                    stack.append(other)

    return caught
